[c10d][ez] Remove goto in PGNCCL and make linter happy for PGNCCL and NCCLUtils

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145855

While working on PGNCCL I found that the code triggers some lint warnings so this PR is to address them or add lint suppressor. 


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[c10d] Remove goto in PGNCCL and make linter happy for PGNCCL and NCCLUtils

[ghstack-poisoned]

Skylion007 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1940,6 +1941,7 @@ void ProcessGroupNCCL::DesyncDebugger::init(
     c10::intrusive_ptr<Store> store) {
   rank_ = rank;
   size_ = size;
+  // NOLINTNEXTLINE(performance-unnecessary-value-param)
   store_ = store;
```
```suggestion
  store_ = std::move(store);

c-p-i-o commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -4634,6 +4636,7 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::barrier(const BarrierOptions& opts) {
   } else if (getBoundDeviceId()) {
     // 2nd choice: Use the bound GPU device id if available.
     // Bounded device id can be passed to `init_process_group`.
+    // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
     barDevIdx = (*getBoundDeviceId()).index();
   } else if (!usedDeviceIdxs_.empty()) {
```
Do you think the following is better?

```suggestion
  } else if (getBoundDeviceId().has_value()) {
    // 2nd choice: Use the bound GPU device id if available.
    // Bounded device id can be passed to `init_process_group`.
    barDevIdx = getBoundDeviceId().value().index();
  } else if (!usedDeviceIdxs_.empty()) {
```

fduwjj replied:
sure, I originally don't want to touch the original code. Yeah we can do that.

fduwjj replied:
interesting, even doing this way, the linter still complains.. I will keep the warning.

fduwjj committed:
Update on "[c10d][ez] Remove goto in PGNCCL and make linter happy for PGNCCL and NCCLUtils"


While working on PGNCCL I found that the code triggers some lint warnings so this PR is to address them or add lint suppressor. 


cc H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
